### DISCUSSION TITLE
Build system

### DISCUSSION THREAD

**Discussion by kristojorg** [readium/ts-toolkit#6] [Ideas]
Hi all,

As discussed in the weekly meeting yesterday, here is my proposal for a build system for Readium web. 

# Our needs

1. **Transpile** Typescript into JS with properly exported types
2. **Bundle** into various module formats for consumers to import: CommonJS, ESM, and UMD 
3. **Watch Mode** for development
4. **Monorepo** support
5. **Testing** setup for unit and possibly integration tests
6. **Treeshaking** for size optimization
7. **Code quality** choices and enforcement
8. **Examples** should live in this repo and be easy to start working with 
9. **CI/CD** - code should be linted, built and tested on every PR. Example would ideally be deployed somewhere so newcomers can easily demo.

# The recommendation

As you probably know, the world of web build systems can be quite frustrating. There is seemingly endless configuration, difficult debugging, and a rapidly changing ecosystem of tools and best practices. It's prettymuch the embodiment of why web dev is annoying and difficult. Over the past 5 years, I've been reaching more and more towards so-called "zero-config" tools, which basically set up and maintain a ready-to-use build and development environment for you with best practices. Usually they can be configured, but don't need to be. For TS libraries I recommend [TSDX](https://tsdx.io). It's extremely simple, has all the features we need, and is officially recommended by Typescript themselves. 

## [TSDX](https://tsdx.io)

You can read more about it on the website, but below is an overview of what it uses internally. The nice part is you don't really need to know most of this because it is already configured and for our use case there would be little need to change anything:

- [Rollup](https://rollupjs.org/guide/en/) for transpilation and bundling. This has become the defacto bundler for libraries, while Webpack remains the main bundler for full apps. 
- [Jest](https://jestjs.io) for testing. Another defacto test runner and assertion library. Generally has replaced the Mocha/Chai/Sinon combination with a single package.
- [Prettier](https://prettier.io) for formatting. I have found it to be extremely useful. I never think about formatting while writing code anymore (a surprising amount of effort used to go in to making lines the right length and brackets line up et). It also importantly guarantees that code is formatted the same when it reaches Github, which means you won't ever see formatting changes in a git diff, only real code changes.
- [Eslint](https://eslint.org) for linting with a very good set of defaults that can be modified if desired. You don't get formatting based lint errors because that's all handled automatically by Prettier, only actual code-quality stuff.
- [Parcel](https://parceljs.org) for the example playground. Parcel is a simple bundler that is great for a little development playground.
- Github Workflows for CI/CD.

I don't think we would need any configuration changes, besides maybe massaging some formatting or lint settings if desired. I have used TSDX many times and have been very pleased with the results, and it is also in use by many other projects and companies large and small. Actively maintained, official Typescript recommendation, etc.

## Other options

I haven't tried many other zero-config setups (not sure there are many others for this use case?), but we could always set up our own custom build system if that is desired. Some tooling options:

- **Transpiling:** Typescript's compiler or Babel
- **Bundling:** Webpack or Rollup, we would want to configure this with tree shaking, watch mode, and bundling to multiple module formats.
- **Testing:** Jest, Mocha, Jasmine, and if it doesn't have an assertion library built in then probably Sinon or another
- **Formatting:** Prettier, basic eslint formatting rules, or no formatter
- **Code Quality:** Eslint (Tslint is deprecated in favor of eslint) or no linting
- **Examples:** Webpack or parcel or just plain HTML if we wanted to.
- **CI/CD:** Github workflows, CircleCI, Travis, or something else.

Let me know what you all think. I'm happy to make a PR adding TSDX if you'd like. Would love to get this set up soon so that I can start using the shared modules!

**Comment by mickael-menu** [comment]
> **Prettier** for formatting. I have found it to be extremely useful. I never think about formatting while writing code anymore (a surprising amount of effort used to go in to making lines the right length and brackets line up et). It also importantly guarantees that code is formatted the same when it reaches Github, which means you won't ever see formatting changes in a git diff, only real code changes.

👍 The anti-bikeshedding auto formatting of Go really grew on me, and that's something I'd like to add on mobile repos as well.

For the rest I'll let people more versed in Web development chime in, but I'll suggest two guidelines:

* Staying with "in-house" GitHub features when they fit the bill, for simplicity (e.g. GitHub workflows and actions).
* Favoring well-established tools instead of cutting-edge, to reduce friction for new/fleeting contributors.

**Reply by kristojorg** [reply]
Yep I agree on all accounts

**Comment by JayPanoz** [comment]
We’ll take a look and discuss it ASAP @ Jellybooks so that we can provide more thorough/robust feedback – not that there is anything surprising, but we have to do our homework about TSDX.

Thanks for the proposal.

**Reply by kristojorg** [reply]
Totally understand that. Feel free to ping me if you have any questions and also I'm happy to add an alternative recommendation if you end up coming up with something different

**Reply by JayPanoz** [reply]
@kristojorg you mentioned having issues with the monorepo aspect after testing TSDX, could you give further details?

CC’ing @julienetie so that it can be easier to have a conversation, options, etc. around that.

**Reply by JayPanoz** [reply]
@kristojorg If I’m understanding correctly, you have to rely on Lerna for the monorepo aspect e.g. [this template from one of the maintainers](https://github.com/jaredpalmer/tsdx-monorepo)? So it adds an extra layer that introduces some uncertainty/complexity?

**Reply by julienetie** [reply]
Hi @kristojorg, 
I've been trying to understand how TSDX can enhance the build setup. Despite being Zero config, It seems like it may be an additional dependency to maintain/ worry about.

Rollup seems to handle most of the requirements (Including the concept of monorepos (as with Webpack too). I think it could be worth sticking to Rollup or Webpack to perform the same desired tasks. But let me know if I'm overlooking any other beneficial aspect of TSDX.

**Reply by kristojorg** [reply]
@JayPanoz you're right that you have to rely on lerna, and the complexity is just that there are some things you have to fiddle with to get everything working, but I know it is possible.

@julienetie the main thing TSDX would provide is abstracting the build, lint and test configs away. That way we just have to update tsdx every once in a while instead of having to manage all of the underlying dependencies ourselves. 

That said, if you prefer, we could just set it all up ourselves. If we were to do that, I would suggest we use Rollup and Yarn workspaces with Lerna.

**Reply by julienetie** [reply]
Thanks @kristojorg that does bring some clarity. 
I don't think it would be a problem to do it ourselves, Rollup, yarn and Lerna sounds cool.

**Reply by kristojorg** [reply]
Ok I'll try to make a PR next week with that setup

**Reply by JayPanoz** [reply]
Alright, so @julienetie has done some research on different options to get a more precise idea of what things could look like going this way i.e. doing it ourselves around Rollup + Yarn + Lerna. This is an attempt to help discuss and design a sensible custom config so he has also tried listing benefits of tools that appear in this list.

**Watcher: Chokidar**
 
- This standalone watcher can be used to conditionally delegate and ignore directory and file changes
- It’s independent from the bundler (if we decide to change the bundler, the same conditional watch system remains intact)
- It’s popular and reliable, typically part of other popular libraries and bundlers.

**Bundler: Rollup**

- Rollup produces a clean JavaScript build. It’s ideal for library builds as it has a small core with an opt-in approach to common features.
- Rollup can be used as a function for more granular customization.
- It handles Treeshaking
- For Monorepo support It handles multiple inputs and targets, as well as multiple configs (like webpack)
- It’s easy to extend configurations conditionally (like webpack)

**Compiler**

- [Sucrase](https://github.com/alangpierce/sucrase)
    - For development builds
    - It’s faster than Babel
- Babel
    - For production builds

**Code Quality: ESLint, Prettier**
 
**Testing: Mocha/ Jest**

**Server: Custom server.js**
    
- A custom server to serve: 	
    - Local development 
    - Local examples
- Can be ran concurrently with the watcher for development
- Websockets or [Livereload](https://github.com/livereload/livereload-js)
    - For developing examples
    - Decouples the reloading aspect from the rest of the build.

**Reply by kristojorg** [reply]
Ok thanks for providing these details. I'll give my opinions below

**Watcher**
I know Chokidar is popular, but Rollup has a watcher built-in, so I'm not too sure what we would need this for, unless you anticipate a complicated watch setup that default rollup watching doesn't handle? I've never personally needed that, so I don't know the use case.

**Bundler**
I think we are on the same page here.

**Compiler**
I am not sure I follow here. Is the suggestion that we use Rollup+babel for production, and Chokidar+sucrase+rollup for dev? I think if we are concerned about dev build speed, [esbuild](https://esbuild.github.io) would be my suggestion (just due to familiarity). However, I haven't had many issues with standard Rollup+babel.

**Server**
I would personally look at [Parcel](http://Parceljs.org) for local examples. I've been using it for my example stuff for a while and love it. Seems much easier than building a custom server. It can also compile+bundle for running examples itself if we plan to use passive compilation (importing sibling folders directly from example folders). 


Otherwise I think I'm in agreement. The other thing to discuss would be changelog creation. I haven't tried many automatic solutions yet, but I know they exist for monorepos. They would take the commit messages and put it together into a changelog. Works especially well if you use something like semantic commits.

**Reply by julienetie** [reply]
> Watcher: "unless you anticipate a complicated watch setup that default rollup watching doesn't handle?"

I agree, if we can avoid it then we should. But if we need more flexibility (e.g. moving multiple files, handling non-ts and non-js builds) it's just JavaScript. Let's stick with rollup watcher for now.

> Bundler: "I think we are on the same page here."

Yes I think so

> Compiler: 
> -  "Is the suggestion that we use Rollup+babel for production, and Chokidar+sucrase+rollup for dev?"

Yes

> - "I think if we are concerned about dev build speed, esbuild would be my suggestion (just due to familiarity). "

I'd take speed over familiarity in this scenario as it collectively saves us all some time. It's probably worth us trying both to see the practical differences.

> Server: "I would personally look at Parcel for local examples. I've been using it for my example stuff for a while and love it. Seems much easier than building a custom server" 

From my understanding, if rollup is building and we're just serving directories and a bunch of files we probably don't need much. A custom server lets us add what we need. If custom is overkill I would prefer to use libs that serve direct concerns like [http-server](https://www.npmjs.com/package/http-server), [live-server](https://www.npmjs.com/package/live-server) or anything that servers a single purpose. I think Parcel would be overkill acting as a server, but I hear you on your experiences with it.

> "The other thing to discuss would be changelog creation."

Yes this will be great, (I have nothing to add here, anything would be a 1up)

**Reply by arhomberg** [reply]
to briefly chime in, but for various reasons a decision needs to made quickly (and should not deviate too much from what we would do anyway) otherwise there will be a LOT of logistic issues; May I suggest we have an agreed approach by Readium call on Wednesday with any "experiments/test" concluded by Wednesday next week?

**Reply by kristojorg** [reply]
Yes good point. From what I understand, the only piece we need to test at this point is the dev setup. These are the options from the discussion above:

- Rollup bundling, babel (within rollup) compiling our packages, custom server serving HTML example files
- Rollup bundling, sucrase compiling our packages, custom server serving HTML example files
- Esbuild bundling and compiling packages, custom server serving HTML example files.
- Parcel bundling, compiling and serving the HTML example files and source code (we import source code from sibling directories directly: `import Publication from "../shared-models"`).

For production we will use Rollup + Babel. So we can already make a PR with that setup. 

One other question is if we use Yarn 1 or Yarn 2. Yarn 2 makes some pretty significant changes. It supposedly comes with a lot of benefits of course, but I haven't used it. I'm guessing Yarn 1 is what we want unless @julienetie has an opinion here?

**Reply by julienetie** [reply]
Great, 
I think the first two should be pretty similar. 
I'm new to Yarn 2. I've just glanced at a few [article](https://shift.infinite.red/yarn-1-vs-yarn-2-vs-npm-a69ccf0229cd#ef4f)s, it seems like sticking to Yarn 1 makes sense, good call.

**Reply by julienetie** [reply]
Hey @kristojorg  

So what's the plan in terms of implementing. Would it be possible to get something out by Friday? 
I'm happy to help with the Rollup builds if needed.

**Reply by kristojorg** [reply]
I was planning to discuss in the readium meeting today. I'm a bit backed up with some other work so probably won't be able to get this done until early next week. If you wanted to make a PR for the rollup builds (or any piece of this) sooner, I'm happy to review it though

**Reply by julienetie** [reply]
Ditto, (this was a question asked to me). I'll make an attempt at Rollup production/ dev and I'll reach to you to make sure I'm not missing out anything.

**Comment by kristojorg** [comment]
Another option I discovered recently: [tsup](https://tsup.egoist.sh) which is powered by [esbuild](https://github.com/evanw/esbuild). It doesn't support as many features out of the box, we would have to build our own testing, examples, code quality etc. But we could pull the defaults of those from tsdx and use the bundler from tsup if desired.

**Comment by JayPanoz** [comment]
I’ll ask Julien to take part in the discussion should he have anything to clarify, discuss or add. 

@mickael-menu BTW is the test-case for Kotlin/Swift available publicly somewhere? I can recall you mentioning you would gladly share it but fails to remember if you ever had to do that “manually.”

**Reply by mickael-menu** [reply]
> @mickael-menu BTW is the test-case for Kotlin/Swift available publicly somewhere? I can recall you mentioning you would gladly share it but fails to remember if you ever had to do that “manually.”

You can find the automated tests for the shared models here:
* https://github.com/readium/r2-shared-kotlin/tree/develop/r2-shared/src/test/java/org/readium/r2/shared/publication
* https://github.com/readium/r2-shared-swift/tree/develop/r2-shared-swiftTests

For the parsers here:
* https://github.com/readium/r2-streamer-kotlin/tree/develop/r2-streamer/src/test/java/org/readium/r2/streamer/parser
* https://github.com/readium/r2-streamer-swift/tree/develop/r2-streamer-swiftTests/Parser

And these are my manual regression test sheets:
* [Kotlin](https://docs.google.com/spreadsheets/d/1febeIllWaLUetEL7WIxHQh0h_qhoc2iCP15r4u_zEt0/edit?usp=sharing)
* [Swift](https://docs.google.com/spreadsheets/d/1NYjU80NHHkQIZE39rOKn4dV7oNhkfwXj6cKtl04-lp4/edit?usp=sharing)
